pkg/processor/build/runtime/java/runtime.go

```diff
@@ -0,0 +1,364 @@
+func (j *java) DetectFunctionHandlers(functionPath string) ([]string, error) {
+	return []string{j.getFunctionHandler()}, nil
+}
+
+// GetProcessorImageObjectPaths returns a map of objects the runtime needs to copy into the processor image
+// the key can be a dir, a file or a url of a file
+// the value is an absolute path into the docker image
+func (j *java) GetProcessorImageObjectPaths() map[string]string {
+	return map[string]string{
+		j.handlerJarPath: path.Join("opt", "nuclio", path.Base(j.handlerJarPath)),
+	}
+}
+
+// GetExtension returns the source extension of the runtime (e.g. .go)
+func (j *java) GetExtension() string {
+	return "jar"
+}
+
+// GetName returns the name of the runtime, including version if applicable
+func (j *java) GetName() string {
+	return "java"
+}
+
+// OnAfterStagingDirCreated will build jar if the source is a Java file
+// It will set generatedJarPath field
+func (j *java) OnAfterStagingDirCreated(stagingDir string) error {
+	buildPath := j.FunctionConfig.Spec.Build.Path
+	j.Logger.DebugWith(
+		"Java OnAfterStagingDirCreated", "buildPath", buildPath, "stagingDir", stagingDir)
+
+	handlerBuildDir := path.Join(stagingDir, "java-handler-build")
+	if err := os.Mkdir(handlerBuildDir, 0777); err != nil {
+		return errors.Wrap(err, "Can't create handler build dir")
+	}
+
+	userJarPath := path.Join(handlerBuildDir, userHandlerJarName)
+	if err := j.buildUserJar(buildPath, userJarPath); err != nil {
+		return err
+	}
+
+	var err error
+	handlerImageName := fmt.Sprintf("nuclio/handler-builder-java-onbuild:%s-%s", j.versionInfo.Label, j.versionInfo.Arch)
+	j.handlerJarPath, err = j.runDockerJavaBuild(handlerBuildDir, handlerImageName)
+
+	return err
+}
+
+func (j *java) getFunctionHandler() string {
+	return "Handler"
+}
+
+func (j *java) GetProcessorBaseImageName() (string, error) {
+	return fmt.Sprintf("nuclio/handler-java:%s-%s",
+		j.versionInfo.Label,
+		j.versionInfo.Arch), nil
+}
+
+// "reverser.jar:Reverser" -> "Reverser"
+func (j *java) handlerClassName(handler string) string {
+	fields := strings.Split(handler, ":")
+	if len(fields) == 1 {
+		return fields[0]
+	}
+	return fields[1]
+}
+
+func (j *java) createBuildFile(stagingBuildDir string) error {
+	buildFilePath := path.Join(stagingBuildDir, buildFileName)
+	if common.IsFile(buildFilePath) {
+		j.Logger.InfoWith("Found user build file, using it", "path", buildFilePath)
+		return nil
+	}
+
+	buildTemplate, err := template.New("build").Parse(buildTemplateCode)
+	if err != nil {
+		return err
+	}
+
+	buildFile, err := os.Create(buildFilePath)
+	if err != nil {
+		return err
+	}
+
+	defer buildFile.Close()
+
+	data := map[string]interface{}{
+		"Dependencies": j.FunctionConfig.Spec.Build.Dependencies,
+		"Handler":      j.handlerClassName(j.FunctionConfig.Spec.Handler),
+	}
+	return buildTemplate.Execute(buildFile, data)
+}
+
+func (j *java) createDockerFile(dockerfilePath, imageName string) error {
```

Comment:
Please refrain from `(x, y <type>)` and opt for `(x <type>, y <type>)` as per the convention. The reason being that this encourages unnatural ordering of arguments

Reply:
OK